### PR TITLE
Features/tasks/7947 update cache record and definition

### PR DESCRIPTION
**Pull request by laum18**
PivotTable row fields are updated based on source data table.

### CODE REVIEW

File: EPPlus/ExcelWorksheet.cs [hunk 1]
```diff
@@ -4010,6 +4010,8 @@ private void SavePivotTables()
 					pt.CacheDefinition.SetXmlNodeString(ExcelPivotCacheDefinition.SourceAddressPath, address);
 				}
 
+
+				// TODO: Move all CacheDefinition saving to workbook.save(). This is saving repeatedly for each worksheet.
```
Comment (shawncrabtree) [inline EPPlus/ExcelWorksheet.cs]: Add this as a task and put the task number in the TODO

File: EPPlus/Table/PivotTable/ExcelPivotTable.cs [hunk 2]
```diff
@@ -941,42 +946,112 @@ internal int CacheID
...
+			{
+				if (!string.IsNullOrEmpty(rowItem.ItemType))
+				{
+					this.WorkSheet.Cells[currentDataRow++, startDataCol].Value = "Grand Total";
+					continue;
+				}
+
+				var fieldIndex = this.RowFields[rowItem.RepeatedItemsCount].Index;
+
+				// TODO: Account for grouped rowItems.
```
Comment (shawncrabtree) [inline EPPlus/Table/PivotTable/ExcelPivotTable.cs]: Task number

File: EPPlus/Table/PivotTable/CacheRecordNode.cs [hunk 3]
```diff
@@ -201,6 +196,30 @@ public void Update(IEnumerable<object> row, ExcelPivotCacheDefinition cacheDefin
 				col++;
 			}
 		}
+
```
Comment (EvanSchallerer) [inline EPPlus/Table/PivotTable/CacheRecordNode.cs]: Extra newline

File: EPPlus/Table/PivotTable/ExcelPivotCacheRecords.cs [hunk 4]
```diff
@@ -121,16 +140,113 @@ public ExcelPivotCacheRecords(XmlNamespaceManager ns, XmlDocument cacheRecordsXm
...
+			this.Count = this.Records.Count;
+		}
+		#endregion
+
+		#region Public Methods
+		/// <summary>
+		/// Checks if a given row item exists.
+		/// </summary>
+		/// <param name="nodeIndices">A list of tuples containing the pivotField index and item value.</param>
+		/// <returns></returns>
```
Comment (EvanSchallerer) [inline EPPlus/Table/PivotTable/ExcelPivotCacheRecords.cs]: Returns comment

File: EPPlus/Table/PivotTable/ExcelPivotTable.cs (hunk 2 above)
Comment (EvanSchallerer) [inline EPPlus/Table/PivotTable/ExcelPivotTable.cs]: Task

File: EPPlus/Utils/XmlCollectionBase.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml;
+
+namespace OfficeOpenXml.Utils
+{
+	/// <summary>
+	/// The base class for xml collections.
+	/// </summary>
+	/// <typeparam name="T"></typeparam>
```
Comment (EvanSchallerer) [inline EPPlus/Utils/XmlCollectionBase.cs]: Comment

File: EPPlusTest/Table/PivotTable/ExcelPivotCacheDefinitionTest.cs [hunk 6]
```diff
@@ -7,11 +33,212 @@
...
+				this.AssertCacheRecord(cacheRecords, 0, 1, PivotCacheRecordType.x, "0");
+				this.AssertCacheRecord(cacheRecords, 1, 1, PivotCacheRecordType.x, "3");
+				this.AssertCacheRecord(cacheRecords, 2, 1, PivotCacheRecordType.x, "2");
+				this.AssertCacheRecord(cacheRecords, 0, 2, PivotCacheRecordType.x, "0");
+				this.AssertCacheRecord(cacheRecords, 1, 2, PivotCacheRecordType.x, "2");
+				this.AssertCacheRecord(cacheRecords, 2, 2, PivotCacheRecordType.x, "0");
+				this.AssertCacheRecord(cacheRecords, 0, 3, PivotCacheRecordType.n, "100");
+				this.AssertCacheRecord(cacheRecords, 1, 3, PivotCacheRecordType.n, "90000");
+				this.AssertCacheRecord(cacheRecords, 2, 3, PivotCacheRecordType.n, "10");
+				//TODO: Test refreshing all pivot tables.
```
Comment (EvanSchallerer) [inline EPPlusTest/Table/PivotTable/ExcelPivotCacheDefinitionTest.cs]: I think we can remove this todo

File: EPPlusTest/Table/PivotTable/ExcelPivotTableTest.cs [hunk 7]
```diff
@@ -51,5 +77,151 @@ public void PivotTableXmlLoadsCorrectly()
...
+				var pivotField2 = pivotTable.Fields[1];
+				Assert.AreEqual(4, pivotField2.Items.Count);
+				var pivotField3 = pivotTable.Fields[2];
+				Assert.AreEqual(6, pivotField3.Items.Count);
+				this.CheckFieldItems(pivotField3);
+				var pivotField4 = pivotTable.Fields[3];
+				Assert.AreEqual(0, pivotField4.Items.Count);
+				Assert.AreEqual(7, pivotTable.RowItems.Count);
+				Assert.AreEqual("Blue", worksheet.Cells[11, 9].Value);
+				//Assert.AreEqual(100, worksheet.Cells[11, 10].Value);
```
Comment (EvanSchallerer) [inline EPPlusTest/Table/PivotTable/ExcelPivotTableTest.cs]: We might want a todo here explaining the commented out lines and link them to the relevant task.

File: EPPlusTest/Table/PivotTable/CacheItemTest.cs [hunk 8]
```diff
@@ -37,90 +37,66 @@ public class CacheItemTest : PivotTableTestBase
 	{
 		#region Constructor Tests
 		[TestMethod]
-		public void CacheItemParsesTypeCorrectly()
+		public void CacheItemWihtoutGivenValue()
```
Comment (zpfaltersack) [inline EPPlusTest/Table/PivotTable/CacheItemTest.cs]: typo

File: EPPlus/Table/PivotTable/ExcelPivotCacheRecords.cs (hunk 4 above)
Comment (rfoushee) [inline EPPlus/Table/PivotTable/ExcelPivotCacheRecords.cs]: missing returns

**Pull request merged**